### PR TITLE
Products wireup and delete functionality

### PR DESCRIPTION
**Pull request by EdiongAlexander**
## Description
[Please put a description of the changes this branch introduces here.]

## To Test
[Please write instructions for testing your changes locally here. Include any data setup, tests, urls, and potential snags in local testing.]

## Checklist
- [ ] I have looked over the diffs.
- [ ] There are no errors/failing tests from my changes on this branch.
- [ ] I have requested a review on this PR and re-assigned the Jira ticket.

## Screenshots
[Any screenshots of the changes can go here.]

## Extra Notes
[Please include any extra notes you may have about the changes here.]

### CODE REVIEW

File: src/components/products/index.js [hunk 1]
```diff
@@ -20,10 +22,23 @@ const ProductsDashboard = () => {
...
+    const dispatch = useDispatch();
+
+    const token = getTokenFromCookie();
+    const user = useSelector((state) => state.user);
+    const storeData = useSelector((state) => state.storeData);
+    const productsData = useSelector((state) => state.productsData);
+
+    useEffect(() => {
+
+        dispatch(getStoreData(user.id, token));
```
Comment (ObiTam) [inline src/components/products/index.js:34]: all data calls should be done from the layout.

File: src/components/products/index.js [hunk 2]
```diff
@@ -20,10 +22,23 @@ const ProductsDashboard = () => {
...
+        dispatch(getProductData(storeData.id, token));
+
+    }, [])
+
 
     // from usePagination hook
-    let data = useFilter("products", activeTab, PRODUCT_DATA, searchTerm, filterObject).filteredData;
+    let data = useFilter("products", activeTab, productsData, searchTerm, filterObject).filteredData;
```
Comment (ObiTam) [inline src/components/products/index.js:41]: We can handle this in a different ticket but I want this rebuilt if possible. it is not readable.

File: src/components/products/index.js [hunk 3]
```diff
@@ -42,17 +57,17 @@ const ProductsDashboard = () => {
     //search and filter modal
     const handleSearch = (searchWord) => setSearchTerm(searchWord)
     const handleFilterObject = (filterObject) => setFilterObject(filterObject)
-    const goToEdit = (sku) => navigate(`/products/edit/${sku}`)
-
 
+    const goToEdit = (productID) => navigate(`/products/edit/${productID}`)
```
Comment (ObiTam) [inline src/components/products/index.js:61]: lets use sku so we have more readable urls

File: src/redux/action.js [hunk 4]
```diff
@@ -91,6 +114,25 @@ export const discardDraft = (hash) => dispatch => {
     ...hash,
   })
 }
+export const deleteProductByID = (hash) => dispatch => {
```
Comment (ObiTam) [inline src/redux/action.js:117]: we don't need to do this. instead make a call to the backend to update the products after a product is deleted.

File: src/redux/action.js [hunk 5]
```diff
@@ -91,6 +114,25 @@ export const discardDraft = (hash) => dispatch => {
...
+  dispatch({
+    type: 'DELETE_PRODUCT',
+    ...hash,
+  })
+}
+
+export const getStoreData = (userID, token) => async (dispatch) => {
```
Comment (ObiTam) [inline src/redux/action.js:124]: make sure userID is not null

File: src/redux/action.js [hunk 6]
```diff
@@ -91,6 +114,25 @@ export const discardDraft = (hash) => dispatch => {
...
+  await fetchData(dispatch, `stores/${userID}`, 'storeData', token);
+}
+
+export const getProductData = (storeID, token) => async (dispatch) => {
```
Comment (ObiTam) [inline src/redux/action.js:128]: make sure storeID is not null

**Comment by ObiTam** [comment]
image
these have to all be the same size

**Comment by ObiTam** [comment]
image
search isn't working

**Comment by ObiTam** [comment]
image
pagination should only show when we have more items than can be viewed all at once. lets say greater than 10 items

**Comment by ObiTam** [comment]
what happens if I go to a page

/products/edit/:id or /products/edit/:sku

and :id or :sku is not a valid sku

**Comment by ObiTam** [comment]
please handle the conflict @EdiongAlexander

File: src/components/addProduct/helpers/imageDisplay.js [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
 export const ImageDisplay = ({ selectedFiles, onDelete }) => {
     return (
       <div className="flex flex-wrap gap-8 pt-4">
-        {selectedFiles.map((file, index) => (
+        {selectedFiles?.map((file, index) => (
           <div className="flex gap-[12px]" key={index}>
             <div key={index} className="mb-4 relative">
               <img
-                src={file.url}
+                src={file.url.startsWith("/uploads") ? renderValidUrl(file.url) : file.url}
```
[the remark this answers is not in the document]
Comment (EdiongAlexander) [reply src/components/addProduct/helpers/imageDisplay.js:11]: Doesn't work with file upload for now

**Comment by ObiTam** [comment]
const deleteItem = (e, name) => {
    if (name === "products") {
      dispatch(deleteProduct(id, token));

      setTimeout(() => {
        dispatch(getProductData(storeData.id, token));
      }, [2000])
    } else {
      e?.stopPropagation();
      const newUsers = roles.filter((r) => r.id !== user.id);
      dispatch(updateUserRole({ roles: newUsers }));
      setOpenDeleteModal(false);
      setShowDetails(false);
    }
  }
there is nothing to show that the product has been deleted (toast). 
Instead of using dispatch to redux you can call the deleteRequest function directly with an await command then get products after. 
It needs to be that when a product is deleted and the toast is thrown that it was successful then the product immediately dissapears from the list.

**Review by ObiTam — APPROVED** [review]
Looks Perfect to me.

I pushed my fixes for the slight changes. 

@Eshy10 please view.

**Comment by ObiTam** [comment]
Hi @EdiongAlexander,

can we add a loading state for the dashboard. There's a flicker i'm not comfortable with.
See below initialize ordersData as null.
image

then you add a loading state when ordersData is null. 
image

So the data will not flicker with the empty array state before displaying after update.

**Comment by Eshy10** [comment]
@EdiongAlexander i am stuck on loading with this error on production via the preview url (if this error is occurring here, it will definitely occur in production)

Screenshot 2024-04-11 at 12 04 13
Screenshot 2024-04-11 at 12 06 18

**Comment by ObiTam** [comment]
@Eshy10 

The issue has been fixed. 
There's still a flicker on the main overview page. this will be fixed in Blossom's PR because it is on store initialization. 

Please review.

cc @EdiongAlexander  @Blossom1-chike

**Comment by Eshy10** [comment]
> Visit Preview

@Blossom1-chike please fix this in your current pr

**Pull request merged**